### DISCUSSION TITLE
Converting the Erfurt Predigerkirche Demo sampleset to GrandOrgue.

### DISCUSSION THREAD

**Discussion by oleg68** [GrandOrgue/OdfEdit#11] [General]
I'm trying to convert the [Erfurt Predigerkirche Demo sampleset from Piotr Grabowski](https://piotrgrabowski.pl/erfurt-predigerkirche/) ODF to Grandorgue with OdfEdit 2.1.

After opening the hauptwerk xml file and building the object tree, I can not see any Rank object in the organ.

**Comment by eturpault** [comment]
Yes I already tried to convert this sample set as well and it doesn't give good results. I plan to work on it later, it is described in a way very different from older sample sets of Piotr Grabowski. There are no Rank objects because OdfEdit doesn't detect properly the Stops objects.

**Comment by eturpault** [comment]
I analyzed this Erfurt Predigerkirche Demo sample set for Hauptwerk. In fact there are missing objects (called StopRank) to link Stop to Rank objects, so since there are no ranks associated to stops, OdfEdit doesn't create the rank objects. This doesn't seem to be normal, any other sample set that I tried have StopRank objects defined, even in the demo sample sets for the allowed stops. I contacted Piotr Grabowski to ask him why.

**Comment by oleg68** [comment]
Hi, @eturpault. I've just tried conversion of the Erfurt Predigerkirche Demo sampleset to GrandOrgue. Now all stops are here, but only Distant ranks are on file. Both Front and Rear are absent.

**Comment by eturpault** [comment]
Indeed there is this limitation. In the recent demo sample sets of Piotr Grabowski that I know (Erfurt, Düren) the link between the stops and the ranks is not done as per the specification (i.e. using an intermediate StopRank object) as it is the case in older sample sets of Piotr and in all the one of Sonus Paradisi and Augustine that I know, and which permits to have sereval Ranks linked to one Stop. 
Piotr uses for Erfut and Düren an attribute called Hint_PrimaryAssociatedRankID of the Stop objet which permits to refer to one rank only, and which should be only a hint, not a functional link. With this way to design the sample set, in OdfEdit I can only do a workaround by using this hint attribute to link one rank (the close one in the case of these sample set) per stop. The ranks of the other channels (front, rear) are defined in the HW ODF but they are linked to none Stop (and none StopRank since none StopRank object is defined in these ODF), and the samples defined inside these ranks are in fact the one used by the close ranks. Inside the files of the sample set, one can see that there are pipe samples defined for a single channel, not for three channels.
So my assumption is that Piotr intentionally bridles his recent demo sample sets by offering only one channel and using a simple link between Stops and Ranks. He didn't answered to my mail so I have not got confirmation of my assumption.
In the description page of these sample sets Piotr writes : 
> The only limitation of demo versions is the number of available stops.

This is not completely true, he should rather write : 
> The only limitation of demo versions is the number of available stops **and channels**.

**Reply by oleg68** [reply]
> Inside the files of the sample set, one can see that there are pipe samples defined for a single channel, not for three channels.

There are the samples for all six channels (3 stereo pairs):

A0_D
A0_F
A0_R

and appropriate folders for releases as well.

> The ranks of the other channels (front, rear) are defined in the HW ODF but they are linked to none Stop

Could you in this case create the Rank GO ODF objects without linking to any stops? I could link them manually later using the new feature from OdfEdit 2.3.beta?

**Comment by eturpault** [comment]
You are totally right, I don't know how I can have missed these _D _F _R folders...
Yes I can create ranks in GO ODF corresponding to HW Ranks (with pipe samples inside, not with noise samples) not linked to a GO Rank. Then as you say they can be linked manually to stops in OdfEdit thanks to the new Parents or Children button.
I have just implemented it, it seems to work well. I will share a new beta release for Linux (I have first to solve an issue with my Linux virtual machine).

**Comment by oleg68** [comment]
After testing ODFEdit 2.3 with this samplesets I noted, that the compass etstension in this sampleset is wrong: there is lack of Pipe999PitchTuning lines. For example:

```
Pipe028Gain=1.5
Pipe028MIDIKeyNumber=63
Pipe028HarmonicNumber=16
Pipe028LoadRelease=N
Pipe028=..\OrganInstallationPackages\002524\P Pommer 4\A0_D\063-d#.wav
Pipe028ReleaseCount=4
Pipe028Release001=..\OrganInstallationPackages\002524\P Pommer 4\R0_D\063-d#.wav
Pipe028Release001MaxKeyPressTime=-1
Pipe028Release002=..\OrganInstallationPackages\002524\P Pommer 4\R1_D\063-d#.wav
Pipe028Release002MaxKeyPressTime=154
Pipe028Release003=..\OrganInstallationPackages\002524\P Pommer 4\R2_D\063-d#.wav
Pipe028Release003MaxKeyPressTime=375
Pipe028Release004=..\OrganInstallationPackages\002524\P Pommer 4\R3_D\063-d#.wav
Pipe028Release004MaxKeyPressTime=438
Pipe029Gain=1.5
Pipe029MIDIKeyNumber=64
Pipe029HarmonicNumber=16
Pipe029LoadRelease=N
Pipe029=..\OrganInstallationPackages\002524\P Pommer 4\A0_D\062-d.wav
Pipe029ReleaseCount=4
Pipe029Release001=..\OrganInstallationPackages\002524\P Pommer 4\R0_D\062-d.wav
Pipe029Release001MaxKeyPressTime=-1
Pipe029Release002=..\OrganInstallationPackages\002524\P Pommer 4\R1_D\062-d.wav
Pipe029Release002MaxKeyPressTime=223
Pipe029Release003=..\OrganInstallationPackages\002524\P Pommer 4\R2_D\062-d.wav
Pipe029Release003MaxKeyPressTime=342
Pipe029Release004=..\OrganInstallationPackages\002524\P Pommer 4\R3_D\062-d.wav
Pipe029Release004MaxKeyPressTime=504
```

As you can see ``Pipe029PitchTuning=200`` is necessary, but it is missing.

**Reply by eturpault** [reply]
The compass extension feature is not yet implemented in OdfEdit, I have to work on it.
What you observe in the GO ODF with this sample set is a copy-paste made by OdfEdit from what is defined in the HW ODF. I double-checked and indeed the attack/releases samples "062-d.wav" of the stop "P Pommer 4" are actually mapped to the pipe which the MIDI note is 64 ! They are mapped also to the pipe which the MIDI note is 62 (normal) and 66 !
I observed such wrong pipe assignment with other demo sample sets of Piotr Grabowski (Ernelo and Alessandria for example), maybe it is intended to limit the experience with the demo sample sets. I never observed such issue with the demo sample sets of Sonus Paradisi.

**Reply by oleg68** [reply]
> "062-d.wav" of the stop "P Pommer 4" are actually mapped to the pipe which the MIDI note is 64 ! They are mapped also to the pipe which the MIDI note is 62 (normal) and 66 !

It would be all right with Pipe029PitchTuning=200 and Pipe031PitchTuning=400. Maybe there is an an equivalent of Pipe099PitchTuning of GO in the HW format?

**Reply by eturpault** [reply]
I compared in detail in the HW ODF the objects which define the pipes with MIDI note 62 / 64 / 66. All are sharing the same Sample objects which use the files "062-d.wav", in these **Sample** objects there is the following attribute:
- Pitch_ExactSamplePitch=579.59

At upper level, in the **Pipe** objects of each MIDI note, there is an interresting different attribute value:
- for MIDI note 62 : Pitch_OriginalOrgan_PitchHz=579.59
- for MIDI note 64 : Pitch_OriginalOrgan_PitchHz=651.76357844136
- for MIDI note 66 : Pitch_OriginalOrgan_PitchHz=731.57988127073

So using these attributes it should be possible to define a Pipe999PitchTuning attribute value when there is a pitch value difference between the Pipe and Sample objects. For that, I should convert the two PitchHz values of Pipe and Sample objects in equivalent MIDI notes, then set Pipe999PitchTuning = 100 * (Pipe MIDI note - Sample MIDI note). Do you agree ?
For this convertion, I need to know the tuning reference of the organ (A4=440Hz commonly), I think that I can get it from the following attribute defined in the General object of the sample set (if it is defined, else I will take 440 by default): 
- AudioEngine_BasePitchHz=435

I could use for example the formula described in [this website](https://www.music.mcgill.ca/~gary/307/week1/node28.html) or [this one](https://newt.phys.unsw.edu.au/jw/notes.html). Using loopup table to make fastest convertion may not work with organs which have unequal temperament or which are not tuned with A4=440Hz.

**Reply by eturpault** [reply]
So there is no willing of Piotr Grabowski to detune his demo sample set, this is fine !

**Reply by eturpault** [reply]
I implemented the pipes pitch tuning as I described in my previous comment and it works well : the attribute Pipe999PitchTuning is added in the ranks of the GO ODF when necessary to align the sample frequency to the pipe frequency based on the pitch attributes of the HW ODF. It will be available in the next OdfEdit release.
The sample sets of Piotr Grabowski are using often sample reuse (for example 77 times for Erfurt demo, 288 times for Ermelo demo).
Thanks @oleg68 for your tests and suggestions which permit to improve OdfEdit.

**Reply by oleg68** [reply]
Also I'd remove AcceptRetune=N, because temperaments do not work with it.

**Reply by eturpault** [reply]
I put in the ranks the attribute AcceptsRetuning=N as a workaround in order to avoid a lot of warning/error logs while loading in GrandOrgue the converted sample sets.
You can do the test with Erfurt, if you remove this attribute in the ranks of the ODF, you will get a lot of following log messages in GrandOrgue when loading the ODF:
`Warning: .... no pitch information provided`
`Warning: .... temperament would retune pipe by more than xxx cent`
`Error: .... temperament would retune pipe by xxxx - more than xxx cent`

Same result for example with Smecno of Sonus Paradisi.
According to you what is causing these logs messages and how to avoid them without setting AcceptsRetuning=N ?

**Reply by larspalo** [reply]
AcceptsRetuning=N should only be used for effect stops or noises that are never intended to be re-tuned.

The warnings are simply because no pitch info is embedded in the samples (dwMIDIUnityNote and dwMIDIPitchFraction of the smpl chunk), and/or that no relevant pitch info is supplied in the odf. What you need to do, if the samples lack embedded pitch info, is to convert the pitch described in the hw odf to the corresponding way pitch is described in GO odf - by using Pipe999MIDIKeyNumber= and Pipe999MIDIPitchFraction=.

This is how I did it in LoopAuditioneer to convert from pitch in Hz to note and cent:

```
            int midi_note = (69 + 12 * (log10(givenPitchInHz / 440.0) / log10(2)));
            double midi_note_pitch = 440.0 * pow(2, ((double)(midi_note - 69) / 12.0));
            double cent_deviation = 1200 * (log10(givenPitchInHz / midi_note_pitch) / log10(2));
```

then you have the midi_note as the Pipe999MIDIKeyNumber and the cent_deviation as the Pipe999MIDIPitchFraction.

**Reply by oleg68** [reply]
> `Warning: .... no pitch information provided` 
> 
> Same result for example with Smecno of Sonus Paradisi.

Smecno samples **contain** a pitch information. We need to check more with this sample set without AcceptRetune. Maybe there is some another issue that causes that warning.

**Reply by oleg68** [reply]
> What you need to do, if the samples lack embedded pitch info, is to convert the pitch described in the hw odf to the corresponding way pitch is described in GO odf - by using Pipe999MIDIKeyNumber= and Pipe999MIDIPitchFraction=.

Now ODFEdit does not read the sample files and can not know do the samples contain embedded pitch info.

I haven't met any hw sample set without pitch info since ten years ago.

**Reply by oleg68** [reply]
> I should convert the two PitchHz values of Pipe and Sample objects in equivalent MIDI notes, then set Pipe999PitchTuning = 100 * (Pipe MIDI note - Sample MIDI note). Do you agree ?

I'd do that not for all pipes but only for "second" and "third" references to the same pipe.

**Reply by oleg68** [reply]
> * AudioEngine_BasePitchHz=435

It is a good candidate for converting to `PitchCorrection` value at the organ level.

**Reply by eturpault** [reply]
I add the Pipe999PitchTuning attribute only if I see in the HW ODF a difference between the Pitch_ExactSamplePitch attribute of the sample object and the Pitch_OriginalOrgan_PitchHz attribute in the pipe object. I tested it with several sample sets and it seems to work well when playing the stops on the keyboard, the notes are rendered at the expected frequency.
When the attribute AudioEngine_BasePitchHz is defined in the HW ODF at organ level, I use it in the convertion formula (from pitch frequency to MIDI note), else I use 440Hz by default.

**Reply by eturpault** [reply]
It is indeed strange that sample sets from Sonus Paradisi or Piotr Grabowski have no pich info in the wav files. 
If a solution cannot be found in GrandOrgue to avoid these warnings/errors when AcceptsRetuning=N is not set, or I have to keep AcceptsRetuning=N, or I could try the proposal of @larspalo to set Pipe999MIDIKeyNumber and Pipe999MIDIPitchFraction attributes for each pipe, using as input the attribute Pitch_ExactSamplePitch provided for each sample in the HW ODF, but it would add a lot of attributes in the GO ODF and would increase the convertion duration.

**Reply by oleg68** [reply]
> It is indeed strange that sample sets from Sonus Paradisi or Piotr Grabowski have no pich info in the wav files.

Thay **have** pitch info.

**Reply by larspalo** [reply]
And of course the correct HarmonicNumber must be set to avoid warnings (or down to Pipe999HarmonicNumber). And as @oleg68 correctly already has pointed out, if you have specified the Pipe999MIDIKeyNumber you need to take full responsibility of the pitch as it overrides the embedded info.

**Reply by eturpault** [reply]
For the moment I will let AcceptsRetuning=N in the ranks to avoid too much complexity in the ODF (and in the convertion), and setting neither Pipe999MIDIKeyNumber nor Pipe999MIDIPitchFraction.
Having the temperaments not working doesn't seem a big issue for me, the goal is normally to play an organ in its native temperament. I will add a mention to this limitation in the help of OdfEdit.

**Reply by larspalo** [reply]
@eturpault If you just avoid setting Pipe999MIDIKeyNumber and Pipe999MIDIPitchFraction, but do set the HarmonicNumber correctly when the samples contain correct pitch info - then everything should be working and you should not add AcceptsRetuning=N (which should be done only for noise effects).

**Reply by eturpault** [reply]
About the definition of the attribute Pipe999PitchTuning, setting a multiple of 100 to retune a sample to the pipe to which it is attached will not give perfect result on non equal temperament organs. I have modified my formula to apply the exact cent value between the two frequencies (pipe frequency and sample frequency), it should give more accurate tuning.
For example here in this file what it gives for the Erfurt sample set.
[Erfurt pipes pitch tunings.txt](https://github.com/GrandOrgue/ODFEdit/files/11292921/Erfurt.pipes.pitch.tunings.txt)

@larspalo : I like your proposal, but I already set the attribute Pipe999HarmonicNumber for each pipe, using the value which is defined in the HW ODF (called Pitch_Tempered_RankBasePitch64ftHarmonicNum), it is an integer value, and I have the warnings/errors that I mentioned earlier if AcceptsRetuning=N is not added. According to you how should I compute the HarmonicNumber in order to avoid these warnings/errors when I don't add AcceptsRetuning=N ? Thanks.

**Reply by larspalo** [reply]
In the case @oleg68 quotes above, the only thing you should need to do is to drop the Pipe028MIDIKeyNumber=63 and similar lines and it should work as the HarmonicNumber seems to be set correctly.

**Reply by larspalo** [reply]
> About the definition of the attribute Pipe999PitchTuning, setting a multiple of 100 to retune a sample to the pipe to which it is attached will not give perfect result on non equal temperament organs.

PitchTuning in odf affects the "Original Temperament", PitchCorrection will affect "Any other selected temperament". If the embedded pitch info is correct then no pitch correction should be needed at all, but it is available in GO so that the user can fix tuning errors due to faulty embedded pitch info. But you're perfectly right that the simple 100 cent change won't always be the best option!

**Reply by eturpault** [reply]
I removed AcceptsRetuning=N and Pipe999MIDIKeyNumber=xx attributes, I placed HarmonicNumber at rank level instead of pipe level, I set Pipe999PitchTuning when necessary to align the sample frequency to the pipe frequency at which it is linked. With that I have no more warnings/errors generated by GO when loading the Erfurt demo sample set. No issues as well with Alessandria demo, LedzinyStClement, Skrzatusz of Piotr Grabowski.
But I have still few errors/warnings "temperament would retune pipe..." (less than before) with other sample sets of Piotr Grabowski : Düren demo, Ernelo demo and Sonus Paradisi : Buckeburg, Smecno demo, St Omer demo. Here are their ODF (to place in the folder OrganDefinitions of the sample set) if you could test one and understand why there are still those errors/warnings.
[Ermelo (demo).zip](https://github.com/GrandOrgue/ODFEdit/files/11301018/Ermelo.demo.zip)
[Duren (demo).zip](https://github.com/GrandOrgue/ODFEdit/files/11301019/Duren.demo.zip)
[Buckeburg, Janke Organ, Surround Demo.zip](https://github.com/GrandOrgue/ODFEdit/files/11301030/Buckeburg.Janke.Organ.Surround.Demo.zip)
[Smecno Wet & Surround  DEMO.zip](https://github.com/GrandOrgue/ODFEdit/files/11301033/Smecno.Wet.Surround.DEMO.zip)
[St. Omer - Surround DEMO.organ.zip](https://github.com/GrandOrgue/ODFEdit/files/11301038/St.Omer.-.Surround.DEMO.organ.zip)

**Reply by oleg68** [reply]
> I placed HarmonicNumber at rank level instead of pipe level

Different pipes of Mixtura-like stops usually have different Harmonic Numbers.

> But I have still few errors/warnings "temperament would retune pipe..." (less than before) with other sample sets of Piotr Grabowski : Düren demo, Ernelo demo and Sonus Paradisi : Buckeburg, Smecno demo, St Omer demo. Here are their ODF (to place in the folder OrganDefinitions of the sample set) if you could test one and understand why there are still those errors/warnings.

I've tested you attached the Smecno demo odf. All errors and warnings relate to Mixtura and Cimbale pipes. The reason is the same: incorrect HarmonicNumber value.

These are the right values for Mixtura:

```
Pipe001HarmonicNumber=64
Pipe002HarmonicNumber=64
Pipe003HarmonicNumber=64
Pipe004HarmonicNumber=64
Pipe005HarmonicNumber=64
Pipe006HarmonicNumber=64
Pipe007HarmonicNumber=64
Pipe008HarmonicNumber=64
Pipe009HarmonicNumber=64
Pipe010HarmonicNumber=64
Pipe011HarmonicNumber=64
Pipe012HarmonicNumber=64
Pipe013HarmonicNumber=64
Pipe014HarmonicNumber=64
Pipe015HarmonicNumber=64
Pipe016HarmonicNumber=64
Pipe017HarmonicNumber=64
Pipe018HarmonicNumber=64
Pipe019HarmonicNumber=64
Pipe020HarmonicNumber=64
Pipe021HarmonicNumber=64
Pipe022HarmonicNumber=64
Pipe023HarmonicNumber=64
Pipe024HarmonicNumber=64
Pipe025HarmonicNumber=32
Pipe026HarmonicNumber=32
Pipe027HarmonicNumber=32
Pipe028HarmonicNumber=32
Pipe029HarmonicNumber=32
Pipe030HarmonicNumber=32
Pipe031HarmonicNumber=32
Pipe032HarmonicNumber=32
Pipe033HarmonicNumber=32
Pipe034HarmonicNumber=32
Pipe035HarmonicNumber=32
Pipe036HarmonicNumber=32
Pipe037HarmonicNumber=16
Pipe038HarmonicNumber=16
Pipe039HarmonicNumber=16
Pipe040HarmonicNumber=16
Pipe041HarmonicNumber=16
Pipe042HarmonicNumber=16
Pipe043HarmonicNumber=16
Pipe044HarmonicNumber=16
Pipe045HarmonicNumber=16
Pipe046HarmonicNumber=16
Pipe047HarmonicNumber=16
Pipe048HarmonicNumber=16
Pipe049HarmonicNumber=16
Pipe050HarmonicNumber=16
Pipe051HarmonicNumber=16
Pipe052HarmonicNumber=16
Pipe053HarmonicNumber=16
Pipe054HarmonicNumber=16
Pipe055HarmonicNumber=16
Pipe056HarmonicNumber=16
Pipe057HarmonicNumber=16
Pipe058HarmonicNumber=16
Pipe059HarmonicNumber=16
Pipe060HarmonicNumber=16
Pipe061HarmonicNumber=16
```

And for Cimbale:

```
Pipe001HarmonicNumber=128
Pipe002HarmonicNumber=128
Pipe003HarmonicNumber=128
Pipe004HarmonicNumber=128
Pipe005HarmonicNumber=128
Pipe006HarmonicNumber=128
Pipe007HarmonicNumber=128
Pipe008HarmonicNumber=128
Pipe009HarmonicNumber=128
Pipe010HarmonicNumber=128
Pipe011HarmonicNumber=128
Pipe012HarmonicNumber=128
Pipe013HarmonicNumber=128
Pipe014HarmonicNumber=128
Pipe015HarmonicNumber=128
Pipe016HarmonicNumber=128
Pipe017HarmonicNumber=128
Pipe018HarmonicNumber=128
Pipe019HarmonicNumber=128
Pipe020HarmonicNumber=128
Pipe021HarmonicNumber=128
Pipe022HarmonicNumber=128
Pipe023HarmonicNumber=128
Pipe024HarmonicNumber=128
Pipe025HarmonicNumber=64
Pipe026HarmonicNumber=64
Pipe027HarmonicNumber=64
Pipe028HarmonicNumber=64
Pipe029HarmonicNumber=64
Pipe030HarmonicNumber=64
Pipe031HarmonicNumber=64
Pipe032HarmonicNumber=64
Pipe033HarmonicNumber=64
Pipe034HarmonicNumber=64
Pipe035HarmonicNumber=64
Pipe036HarmonicNumber=64
Pipe037HarmonicNumber=32
Pipe038HarmonicNumber=32
Pipe039HarmonicNumber=32
Pipe040HarmonicNumber=32
Pipe041HarmonicNumber=32
Pipe042HarmonicNumber=32
Pipe043HarmonicNumber=32
Pipe044HarmonicNumber=32
Pipe045HarmonicNumber=32
Pipe046HarmonicNumber=32
Pipe047HarmonicNumber=32
Pipe048HarmonicNumber=32
Pipe049HarmonicNumber=32
Pipe050HarmonicNumber=32
Pipe051HarmonicNumber=32
Pipe052HarmonicNumber=32
Pipe053HarmonicNumber=32
Pipe054HarmonicNumber=32
Pipe055HarmonicNumber=32
Pipe056HarmonicNumber=32
Pipe057HarmonicNumber=32
Pipe058HarmonicNumber=32
Pipe059HarmonicNumber=32
Pipe060HarmonicNumber=32
Pipe061HarmonicNumber=32
```

I think the reason of errors in other sample sets is the same.

**Reply by eturpault** [reply]
The HarmonicNumber values that you provided are the one defined in fact in the attribute Pitch_Tempered_RankBasePitch64ftHarmonicNum of the HW sample set and that I used before, so I restored the setting of HarmonicNumber at pipe level based on this attribute and you are right I have no more errors/warnings with the Smecno sample set. Thank you for your advice.
(by the way in the panel "Voices" of Smecno you can see labels using the new property DispImageNum=0)
No error as well with Erfurt demo, Alessandria demo, LedzinyStClement, Skrzatusz of Piotr Grabowski.
No error as well with Buckeburg , St Omer demo, Prytanée demo of Sonus Paradisi.
So there is a very good improvement.

Unfortunately it doesn't solve all the errors/warnings : 
With Düren demo I get a unique log: 
Warning: rank 054. SW  Oboe 8' (close) pipe ..\OrganInstallationPackages\002526\SW Oboe 8\A0_D\088-e.wav: temperament would retune pipe by more than 600 cent
This pipe has PitchTuning=8 like all the other pipes of this rank.
[Duren (demo).zip](https://github.com/GrandOrgue/ODFEdit/files/11304449/Duren.demo.zip)

With Ernelo demo I get 30 logs (more than 600 cent or 1800 cent) with some attack pipes of background stops.
[Ermelo (demo).zip](https://github.com/GrandOrgue/ODFEdit/files/11304457/Ermelo.demo.zip)
[Ermelo_logs.txt](https://github.com/GrandOrgue/ODFEdit/files/11304456/Ermelo_logs.txt)

**Reply by oleg68** [reply]
In the Düren sampleset the pipe

```
Pipe060HarmonicNumber=8
Pipe060LoadRelease=N
Pipe060PitchTuning=708
Pipe060=..\OrganInstallationPackages\002526\SW Oboe 8\A0_D\088-e.wav
```

requires tuning by 708 cents that is more than 600 cents. It causes the warning. It is a problem of the source sampleset rather than ODFEdit.

But from the other side, why have you added Pipe999Tunining=8 almost for all pipes?

**Reply by eturpault** [reply]
I applied the formula that I mentioned in a previous comments last week : 
> I add the Pipe999PitchTuning attribute only if I see in the HW ODF a difference between the Pitch_ExactSamplePitch attribute of the sample object and the Pitch_OriginalOrgan_PitchHz attribute in the pipe object.

> I have modified my formula to apply the exact cent value between the two frequencies (pipe frequency and sample frequency), it should give more accurate tuning.

In the example of the pipe : 
```
Pipe053HarmonicNumber=8
Pipe053LoadRelease=N
Pipe053PitchTuning=8
Pipe053=..\OrganInstallationPackages\002526\SW Oboe 8\A0_D\088-e.wav
```
the HW ODF defines the following frequencies : 
- Pitch_ExactSamplePitch=1327.48
- Pitch_OriginalOrgan_PitchHz=1333.7295311621

which corresponds to a difference of +8 cent.
I could add Pipe999PitchTuning in the ODF only if the result of the formula is higher than 90 for example, but I think that it is better to add it as soon as it different from 0 in order to have a fine tuning. What do you think ?

How do you do to know which amount of tuning a wav file would require ?
If will remove AcceptsRetuning=N in the next versions of OdfEdit for the pipes ranks, so some users may complain that GrandOrgue is generating the error or warning "temperament would retune pipe by more than nnn cent" for some pipes, I have to be prepared to explain them why.

**Reply by oleg68** [reply]
> I could add Pipe999PitchTuning in the ODF only if the result of the formula is higher than 90 for example, but I think that it is better to add it as soon as it different from 0 in order to have a fine tuning. What do you think ?

I suggest to specify Pipe999PitchTuning only for borrowing pipes, bu to substract the value calculated for the borrowed pipe.

For example, for Duren:

```
# Pipe053PitchTuning should not exist
Pipe060PitchTuning=700 # 700=708-8
```

**Reply by oleg68** [reply]
> If will remove AcceptsRetuning=N in the next versions of OdfEdit for the pipes ranks, so some users may complain that GrandOrgue is generating the error or warning "temperament would retune pipe by more than nnn cent" for some pipes, I have to be prepared to explain them why.

1. This warning is displayed only if `Strict ODF check` is selected.
2. Now, retuning a pipe to a more than 600 cents produces this warning. But almost any sample sets with extended compasses usually tune some pies on one octave, ie 1200 cents.

@larspalo what do you think. Should get this warning rid, because lots of existing sample sets raise it?

**Reply by larspalo** [reply]
@oleg68 If the pitch tuning lines also are written then the warning "should" disappear as the difference between what will be adjusted by temperament and what is adjusted for the original temperament will be (should be) less (which is what I understand the warning is calculated from, not the absolute amount of re-tuning done).

As the warning is only enabled for strict odf checking, I don't think it's necessary to remove it. Also I think the warning is somewhat useful as a sample set designer to find out where errors in for instance pitch info can exist so that they might be corrected.

**Reply by eturpault** [reply]
Indeed if I disable the strick ODF checking I get no more of these logs, so this is the simplest solution to avoid them. I will give this tip if someone complains about these logs.

**Reply by oleg68** [reply]
> If the pitch tuning lines also are written then the warning "should" disappear as the difference between what will be adjusted by temperament and what is adjusted for the original temperament will be (should be) less (which is what I understand the warning is calculated from, not the absolute amount of re-tuning done).

Unfortunally, now the warning doesn't take Pipe999PitchTuning into account. It looks like a bug and should be fixed.

**Reply by larspalo** [reply]
@oleg68 Something might have changed since my last experiment with lots of sample extrapolation which was done in late January this year (2023). At that time, when I re-used/stretched samples to other notes, I noticed that I got some warnings due to temperament re-tuning until I also added the Pipe999PitchTuning lines.

I don't know exactly what could/should be considered a bug in this context, but it could be a very good use case for the warning (to continue to exist) to alert if the temperament would change the pitch from the original (temperament) by a very significant amount - which might not be what a sample set designer intends, and could be a signal that an error in the pitch info might be present.

**Comment by oleg68** [comment]
Another problems are related to Pipe999MIDIKeyNumber:

1. There are wrong values in all non 8' ranks: this value must be the actual midi key if the rank was  8'. For example, for any 4' stop Pipe001MIDIKeyNumber should be 48, not 36.
2. If Pipe999MIDIKeyNumber is specifies in ODF then GrandOrgue ignores the PitchFraction value in the sample and expects the correct Pipe999MIDIKeyFraction is also specified

These problems cause incorrect working of GrandOrgue with all not equal temperaments.

I suggest not to write Pipe999MIDIKeyNumber at all when OdfEdit converts from Hauptwerk.

**Reply by eturpault** [reply]
I added the attribute Pipe999MIDIKeyNumber in OdfEdit v2.3, I found it convenient to check if the sample files are correctly assigned to the corresponding pipe999 attributes. But I admit that I am not aware about its effects in GrandOrgue. Based on your explanations, I have indeed to remove it, it will be done in the next release. Thank you for the warning about that.
